### PR TITLE
plan: refine return type, charset and collation for 'get variable' expression

### PR DESCRIPTION
**Pull request by spongedu**
TiDB do not correctly set return-type, charset and collation when executin sql like `select @@xxx`.
For example:
**In MySQL:**
```sql
mysql> select @@sql_mode;
Field   1:  `@@sql_mode`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8_general_ci (33)
Length:     45
Max_length: 15
Decimals:   31
Flags:


+-----------------+
| @@sql_mode      |
+-----------------+
| NO_ZERO_IN_DATE |
+-----------------+
1 row in set (0.01 sec)
```

**In TiDB:**
```sql
tidb> select @@sql_mode;
Field   1:  `@@sql_mode`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       STRING
Collation:  ? (0)
Length:     0
Max_length: 42
Decimals:   31
Flags:


+--------------------------------------------+
| @@sql_mode                                 |
+--------------------------------------------+
| STRICT_TRANS_TABLES,NO_ENGINE_SUBSTITUTION |
+--------------------------------------------+
1 row in set (0.00 sec)

```

### CODE REVIEW

File: plan/expression_rewriter.go [hunk 1]
```diff
@@ -764,7 +765,11 @@ func (er *expressionRewriter) rewriteVariable(v *ast.VariableExpr) {
 		er.err = errors.Trace(err)
 		return
 	}
-	er.ctxStack = append(er.ctxStack, datumToConstant(types.NewStringDatum(val), mysql.TypeString))
+	e := datumToConstant(types.NewStringDatum(val), mysql.TypeString)
+	e.RetType.Tp = mysql.TypeVarString
+	e.RetType.Charset = er.ctx.GetSessionVars().Systems["character_set_connection"]
```
Comment (XuHuaiyu) [inline plan/expression_rewriter.go]: can "character_set_connection" be defined as a constant?

File: plan/expression_rewriter.go [hunk 2]
```diff
@@ -764,7 +765,11 @@ func (er *expressionRewriter) rewriteVariable(v *ast.VariableExpr) {
 		er.err = errors.Trace(err)
 		return
 	}
-	er.ctxStack = append(er.ctxStack, datumToConstant(types.NewStringDatum(val), mysql.TypeString))
+	e := datumToConstant(types.NewStringDatum(val), mysql.TypeString)
```
Comment (zz-jason) [inline plan/expression_rewriter.go]: `e := datumToConstant(types.NewStringDatum(val), mysql.TypeVarString)`, line 769 can be removed

File: plan/expression_rewriter.go (hunk 1 above)
Comment (spongedu) [reply plan/expression_rewriter.go]: ok

File: plan/expression_rewriter.go (hunk 2 above)
Comment (spongedu) [reply plan/expression_rewriter.go]: done

**Pull request merged**